### ISSUE TITLE
Annoying extra output when tests are failing

### ISSUE THREAD

**Issue by Jacalz** [fyne-io/fyne#4978]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same problem, before opening a new one.
- [X] This issue only relates to a single bug. I will open new issues for any other problems.

### Describe the bug

This has been a problem for as long as I can remember but it is time to finally write it down.
When you get a test that fails, you get a lot of other errors that bubble up and make the output more confusing and harder to read. It would be nice if this could be fixed. 

```
2024/06/25 20:28:44 Fyne error:  Missing UpdateCell callback required for List
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/widget/list.go:686
2024/06/25 20:28:44 Fyne error:  Missing UpdateCell callback required for List
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/widget/list.go:686
2024/06/25 20:28:44 Fyne error:  Missing UpdateCell callback required for List
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/widget/list.go:686
2024/06/25 20:28:44 Fyne error:  Missing UpdateCell callback required for List
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/widget/list.go:686
2024/06/25 20:28:44 Fyne error:  Missing UpdateCell callback required for List
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/widget/list.go:686
2024/06/25 20:28:44 Fyne error:  Failed to load image
2024/06/25 20:28:44   Cause: open ../../theme/icons/fyne.png: no such file or directory
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/canvas/image.go:122
2024/06/25 20:28:44 Fyne error:  Failed to load image
2024/06/25 20:28:44   Cause: open ../../theme/icons/fyne.png: no such file or directory
2024/06/25 20:28:44   At: /var/home/jacob/git/fyne/canvas/image.go:122
2024/06/25 20:28:45 Fyne error:  Missing CreateCell callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:792
2024/06/25 20:28:45 Fyne error:  Missing Length callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:1204
2024/06/25 20:28:45 Fyne error:  Missing UpdateCell callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:1268
2024/06/25 20:28:45 Fyne error:  Missing Length callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:1204
2024/06/25 20:28:45 Fyne error:  Missing Length callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:1204
2024/06/25 20:28:45 Fyne error:  Missing UpdateCell callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:1268
2024/06/25 20:28:45 Fyne error:  Missing UpdateCell callback required for Table
2024/06/25 20:28:45   At: /var/home/jacob/git/fyne/widget/table.go:1268
2024/06/25 20:28:48 Fyne error:  Entry cannot wrap single line
2024/06/25 20:28:48   At: /var/home/jacob/git/fyne/widget/entry.go:1467
2024/06/25 20:28:48 Fyne error:  Entry cannot wrap single line
2024/06/25 20:28:48   At: /var/home/jacob/git/fyne/widget/entry.go:1467
2024/06/25 20:28:48 Fyne error:  Entry cannot wrap single line
2024/06/25 20:28:48   At: /var/home/jacob/git/fyne/widget/entry.go:1467
```

### How to reproduce

1. Make a test fail. Something like images maybe?
2. Notice many unrelated errors showing up.

### Example code

Related to tests. See above.

### Fyne version

v2.4.5

### Go compiler version

1.22.4

### Operating system and version

Fedora Silverblue 40

**Comment by cpustejovsky** [comment]
All of these are log messages produced by `LogError` in `log.go`

```go
// LogError reports an error to the command line with the specified err cause,
// if not nil.
// The function also reports basic information about the code location.
func LogError(reason string, err error) {
	log.Println("Fyne error: ", reason)
	if err != nil {
		log.Println("  Cause:", err)
	}

	_, file, line, ok := runtime.Caller(1)
	if ok {
		log.Printf("  At: %s:%d", file, line)
	}
}
```

It's found in 200 locations across the codebase (10 times in the fyne package, 190 times in other packages).

**Comment by Jacalz** [comment]
Absolutely, there is nothing strange there. The problem is why we are seeing all this printouts for failing tests. Are there tests that actually do print those errors (i.e. are incorrect in some sense) but we never see them because testing hides stdout for passing tests?

**Comment by cpustejovsky** [comment]
I'm not sure this is the case.

If you run `go test -v ./theme -run EmptyTheme` and `go test -v ./theme -run JSON` you'll find two tests that are both passing and logging errors.

I checked `TestEmptyTheme` and the logs are recording real issues that I am assuming are results of it being an empty theme. 

I am new to this codebase and have just begun looking at issues.

I may be missing something, but the only fix I'm seeing for this is to suppress the logs in the test by changing the output for log.

**Comment by Jacalz** [comment]
I don't think that invalidates what I wrote in my previous comment. If the tests are passing but logging errors (which we do not see because output normally isn't printed) then that should be fixed in those tests. The tests seem to be doing something wrong or in a bad way.

> I may be missing something, but the only fix I'm seeing for this is to suppress the logs in the test by changing the output for log.

That would only hide the problem, not fix it. We should update the tests that log errors to make sure they don't cause errors to be logged in the first place.

**Comment by cpustejovsky** [comment]
`TestEmptyTheme` is causing errors to be logged around the theme returning `nil` for color, font, and icon.

This is the expected behavior of the `emptyTheme` being passed in for the test.

What would be the solution here and for other places where we're testing the unhappy path?

**Comment by Jacalz** [comment]
Hmm, that's a good question. I think we might want to apply the idea from https://stackoverflow.com/questions/31663229/how-can-stdout-be-captured-or-suppressed-for-golang-testing#31663374 to swallow stdout messages in the specific tests. I don't see any way around that logging.

**Comment by cpustejovsky** [comment]
I agree! That's a good idea.

I'll catalog the passing tests that have error logs and then add code to swallow the the stdout messages for those tests in a PR. Does that work as at least a first step?

**Comment by Jacalz** [comment]
That sounds perfect. I think there is a big win to both making failing test output more readable and catch potentially faulty tests in the process. Thanks for looking at this :)

**Comment by cpustejovsky** [comment]
I've attached the test out put to just have the `LogError`s

Should I contact the developer who wrote these tests when I make a commit to swallow the log output?

I want to make sure these log messages were known about at the time before we hide them.

[errLogs.txt](https://github.com/user-attachments/files/20970581/errLogs.txt)

**Comment by andydotxyz** [comment]
> Should I contact the developer who wrote these tests when I make a commit to swallow the log output?

Probably best to open a PR with what you think is correct and discuss it here. If they are still active contributors feel free to put a mention in the PR, but we accept that people come and go from the project so don't want to make the assumption that having written the code means they need to be available a load of time later.

I think the review process should bring up any uncertainty or ambiguity and we can discuss there.

**Comment by redawl** [comment]
Is it not possible to fix the errors that are being logged, instead of suppressing them? I'd be worried about missing important regressions if we suppress these errors

**Comment by Jacalz** [comment]
Indeed. We can't just blindly hide outputs; that wasn't my intention. Only tests which are testing the error path (and thus causing the log) should hide output.
